zkvm/src/types.rs

```diff
@@ -322,12 +354,61 @@ impl Value {
 }
 
 impl Expression {
-    pub fn constant<S: Into<Scalar>>(a: S) -> Self {
-        let a: Scalar = a.into();
+    pub fn constant<S: Into<ScalarWitness>>(a: S) -> Self {
+        Expression::Constant(a.into())
+    }
+}
+
+impl Neg for Expression {
+    type Output = Expression;
 
-        Expression {
-            terms: vec![(r1cs::Variable::One(), a)],
-            assignment: Some(ScalarWitness::Scalar(a)),
+    fn neg(self) -> Expression {
+        match self {
+            Expression::Constant(a) => Expression::Constant(-a),
+            Expression::Terms(mut terms, assignment) => {
+                for (_, n) in terms.iter_mut() {
+                    *n = -*n;
+                }
+
+                let x = match assignment {
+                    Some(a) => Some(-a),
+                    None => None,
+                };
+                Expression::Terms(terms, x)
```

Comment:
this one-liner would remove the need for the match statement:
```suggestion
                Expression::Terms(terms, assignment.map(|a| -a))
```